File: results/RQ1/05_baseline_qualification/units/p09_qr_code_gen_det/cpp/attempts/0001_seed_1592856389/p09_qr_code_gen_det_cpp/p09_qr_code_gen_det_cpp/src/QRRenderer.cpp
#include "QRRenderer.h"
#include <stdexcept>
#include <string>

std::string QRRenderer::render(const QRMatrix& matrix,
                               const std::string& on,
                               const std::string& off) {
  // Validate markers
  if (on.empty() || off.empty()) {
    throw std::invalid_argument("Markers cannot be empty");
  }

  std::string result;
  result.reserve(QRMatrix::SIZE * (on.length() + off.length()) + QRMatrix::SIZE - 1);
  
  // Render each row
  for (std::size_t row = 0; row < QRMatrix::SIZE; ++row) {
    // Render each column in the current row
    for (std::size_t col = 0; col < QRMatrix::SIZE; ++col) {
      if (matrix.get(row, col)) {
        result += on;
      } else {
        result += off;
      }
    }
    
    // Add newline after each row except the last one
    if (row < QRMatrix::SIZE - 1) {
      result += '\n';
    }
  }
  
  return result;
}

File: results/RQ1/05_baseline_qualification/units/p09_qr_code_gen_det/java/replays/0001_seed_3719859282/repository/src/main/java/QRRenderer.java
/**
 * Provides utilities for rendering QR code matrices as text.
 * Produces a deterministic string representation with exactly one line per matrix row.
 */
public class QRRenderer {
    
    /**
     * Renders a QRMatrix as a text string.
     * Each true cell is represented by '#', each false cell by '.'.
     * There is exactly one line per row in the matrix.
     * 
     * @param matrix the QRMatrix to render
     * @return a string representation of the matrix
     */
    public static String render(QRMatrix matrix) {
        int size = matrix.getSize();
        StringBuilder sb = new StringBuilder();
        
        for (int row = 0; row < size; row++) {
            for (int col = 0; col < size; col++) {
                if (matrix.get(row, col)) {
                    sb.append('#');
                } else {
                    sb.append('.');
                }
            }
            // Add newline after each row except the last one
            if (row < size - 1) {
                sb.append('\n');
            }
        }
        
        return sb.toString();
    }
}